text/3305-static-in-pattern.md

```diff
@@ -0,0 +1,366 @@
+
+Allowing unsafe-to-access statics in patterns (`static mut`s, untrusted `extern` statics, `#[thread_local]` statics) is another possibility. However, I believe this option to be unwise:
+
+- Rust generally has not allowed unsafe operations (like union field accesses) in pattern matches
+- It's not clear where the `unsafe` keyword would go (within the pattern? around the whole `match` or `let`? what about patterns in function parameters?)
+- it requires Rust to commit to and document, and users to understand, when exactly it is allowed to dereference the static when performing a pattern match
+
+The special rules around reachability for static patterns containing references are necessary for soundness. One alternative is to forbid using such types in static patterns completely. However, this alternative expands the set of situations in which adding a reference to a type is a breaking change.
+
+Another alternative is to add a new kind of pattern for runtime equality comparisons, with its own dedicated syntax. In addition to making the language grammar more complex, this option would prevent existing const patterns from being interchangeable with static patterns.
+
+As for not making this change at all, I believe this would be a loss for the language as it would lock out the use-cases described above. This is a very simple feature, it doesn't conflict with any other potential extensions, the behavior and syntax fit well with the rest of the language, and it is immediately understandable to anyone who is already familiar with matching on `const`s.
+
+This feature cannot be fully emulated with a macro, because it's impossible to distinguish a static pattern from a wildcard binding without knowing what statics are in scope. And even an imperfect emulation would potentially require proc macros, which can't easily be used inside the standard library. Also, every crate that needs this feature would need to copy-paste the macro
+implementation into their own source code.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+The D language's switch statement [allows referencing "runtime initialized const or immutable variable[s]"](https://dlang.org/spec/statement.html#switch-statement), which are equivalent to non-`mut` `static`s in Rust, or `const`s in C. However, D also requires that these immutable variables be "of integral type"; they can't contain pointers/references.
+
+```d
+import std;
+
+immutable(int) my_amazing_static = 42;
+
+void main()
+{
+    switch (42)
+    {
+    case my_amazing_static:
+        writeln("Match succeeded");
+        break;
+    default:
+        writeln("Match failed");
+    }
+}
+```
+
+C's `switch` statement does not allow referring to C `const`s.
+
+C++'s `switch` statement allows arbitrary expressions.
```

Comment:
you can't actually use `switch` to match against c++'s equivalent of `static`s:
https://gcc.godbolt.org/z/bTj6nKMYP
```C++
extern const int V;

bool f(int a) {
    switch(a) {
    case V:
        return true;
    default:
        return false;
    }
}
```
```
<source>: In function 'bool f(int)':
<source>:5:10: error: the value of 'V' is not usable in a constant expression
    5 |     case V:
      |          ^
<source>:1:18: note: 'V' was not initialized with a constant expression
    1 | extern const int V;
      |                  ^
Compiler returned: 1
```